consensus/ethash/consensus.go

```diff
@@ -598,12 +598,10 @@ func (ethash *Ethash) Prepare(chain consensus.ChainHeaderReader, header *types.H
 	return nil
 }
 
-// Finalize implements consensus.Engine, accumulating the block and uncle rewards,
-// setting the final state on the header
+// Finalize implements consensus.Engine, accumulating the block and uncle rewards.
 func (ethash *Ethash) Finalize(chain consensus.ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction, uncles []*types.Header, withdrawals []*types.Withdrawal) {
-	// Accumulate any block and uncle rewards and commit the final state root
+	// Accumulate any block and uncle rewards
 	accumulateRewards(chain.Config(), state, header, uncles)
-	header.Root = state.IntermediateRoot(chain.Config().IsEIP158(header.Number))
```

Comment:
I'm a bit unsure about this. 
In blockchain insert, we call state_processor `Process`, which calls `Finalize`: https://github.com/ethereum/go-ethereum/blob/master/core/state_processor.go#L95

I agree that we need to set the `header.Root`, but call here to `IntermediateRoot` also modifies the state (?) by clearing empties. 

I think it still works out, though, because in blockchain, we later call `if err := bc.validator.ValidateState`, which also calls `IntermediateRoot`. So we do it, but a bit later. 

I'm not sure how the meters defined here are affected: https://github.com/ethereum/go-ethereum/blob/master/core/blockchain.go#L1764  . Maybe we have to move something there?

Reply:
> I agree that we need to set the header.Root

I think in block processing, we don't need to mutate header at all. The case is we receive a block generated by block producer, the action here is to apply all transactions inside and then ensure the transition is the same as described by the block.

> which also calls IntermediateRoot. So we do it, but a bit later.

Yes, it's the step for state transition checking. We compute the root hash of local stateDB, ensure the root is same as the one in block header advertised by the block producer.

> I'm not sure how the meters defined here are affected: https://github.com/ethereum/go-ethereum/blob/master/core/blockchain.go#L1764 . Maybe we have to move something there?

Let me check.

Reply:
>> I agree that we need to set the header.Root

> I think in block processing, we don't need to mutate header at all.

Yeah, typo from me. I meant: I agree that we do _not_ need to set the header.Root. So I agred with that part of the change

Reply:
> I'm not sure how the meters defined here are affected: https://github.com/ethereum/go-ethereum/blob/master/core/blockchain.go#L1764 . Maybe we have to move something there?

You are right, I have to fix the metrics.

Reply:
Fixed, I will deploy it on benchmark nodes to see the metrics work or not.